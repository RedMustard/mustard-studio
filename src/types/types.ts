export type KeyboardKeyCount = 88 | 61 | 49 | 37 | 25;
export type KeyboardKeyColor = 'white' | 'black';
export type Octave = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7;
export type OscillatorId = 'osc1' | 'osc2' | 'oscSub';
export type ValueUnit = 'percent' | 'cent' | 'pan';


export type StudioService = {
    settings: {
        master: {
            volume: number,
            pan: number,
        },
        osc1: OscillatorSettings & OscillatorDetune,
        osc2: OscillatorSettings & OscillatorDetune,
        oscSub: OscillatorSettings,
    },
    gainNodes: {
        master: GainNode,
        osc1: GainNode,
        osc2: GainNode,
        oscSub: GainNode,
    },
    panNodes: {
        master: StereoPannerNode,
        osc1: StereoPannerNode,
        osc2: StereoPannerNode,
        oscSub: StereoPannerNode,
    },
};

export type OscillatorSettings = {
    enabled: boolean,
    octave: number,
    pan: number,
    type: OscillatorType,
    volume: number,
};

export type OscillatorDetune = {
    detune: number,
};

export type StudioServiceActionType =
    'SET_MASTER_VOLUME' |
    'SET_MASTER_GAIN' |
    'SET_MASTER_PAN_NODE' |
    'SET_MASTER_PAN_POSITION' |
    'SET_OSC_1_VOLUME' |
    'SET_OSC_2_VOLUME' |
    'SET_OSC_SUB_VOLUME' |
    'SET_OSC_1_GAIN' |
    'SET_OSC_2_GAIN' |
    'SET_OSC_SUB_GAIN' |
    'SET_OSC_1_ENABLED' |
    'SET_OSC_2_ENABLED' |
    'SET_OSC_SUB_ENABLED' |
    'SET_OSC_1_TYPE' |
    'SET_OSC_2_TYPE' |
    'SET_OSC_SUB_TYPE' |
    'SET_OSC_1_DETUNE' |
    'SET_OSC_2_DETUNE' |
    'SET_OSC_1_PAN_NODE' |
    'SET_OSC_2_PAN_NODE' |
    'SET_OSC_SUB_PAN_NODE' |
    'SET_OSC_1_PAN_POSITION' |
    'SET_OSC_2_PAN_POSITION' |
    'SET_OSC_SUB_PAN_POSITION' |
    'SET_OSC_1_OCTAVE' |
    'SET_OSC_2_OCTAVE' |
    'SET_OSC_SUB_OCTAVE';

export type StudioServiceAction = {
    type: StudioServiceActionType;
    payload: any;
};

export type DispatchFunction = (args: StudioServiceAction) => void;

