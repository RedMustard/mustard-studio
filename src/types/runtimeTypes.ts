export enum OscillatorTypeEnum {
    SINE = 'sine',
    SAWTOOTH = 'sawtooth',
    TRIANGLE = 'triangle',
    SQUARE = 'square',
    CUSTOM = 'custom',
}
export const OscillatorTypes = Object.values(OscillatorTypeEnum) as OscillatorType[];

export enum FilterTypeEnum {
    LOWPASS = 'lowpass',
    HIGHPASS = 'highpass',
    BANDPASS = 'bandpass',
    LOWSHELF = 'lowshelf',
    HIGHSHELF = 'highshelf',
    PEAKING = 'peaking',
    NOTCH = 'notch',
    ALLPASS = 'allpass',
}
export const FilterTypes = Object.values(FilterTypeEnum) as BiquadFilterType[];


export enum ValueUnitEnum {
    PERCENT = 'percent',
    CENT = 'cent',
    PAN = 'pan',
    SECONDS = 'seconds',
}

export enum OscillatorIdEnum {
    OSC_1 = 'osc1',
    OSC_2 = 'osc2',
    OSC_SUB = 'oscSub',
}
