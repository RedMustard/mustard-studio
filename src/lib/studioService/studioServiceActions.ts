import { DispatchFunction, OscillatorId } from '../../types/types';
import { stopOscillatorById } from '../oscillators/oscillators';
import { logger } from '../utils/logger/logger';

export const setMasterVolume = (value: number, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_VOLUME', payload: value });
    logger.info(`Master volume set with value ${value * 100}%`);
};

export const setMasterGainNode = (gainNode: GainNode, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_GAIN', payload: gainNode });
    logger.info('Master gain node added', gainNode);
};

export const setMasterPanNode = (panNode: StereoPannerNode, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_PAN_NODE', payload: panNode });
    logger.info('Master pan node added', panNode);
};

export const setMasterPanPosition = (position: number, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_PAN_POSITION', payload: position });
    logger.info('Master pan position set with value ', position);
};

export const setOscillatorVolume = (value: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_VOLUME', payload: value });
            logger.info(`Oscillator 1 volume set with value ${value * 100}%`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_VOLUME', payload: value });
            logger.info(`Oscillator 2 volume set with value ${value * 100}%`);
            break;
        case 'oscSub':
            dispatch({ type: 'SET_OSC_SUB_VOLUME', payload: value });
            logger.info(`Oscillator 2 volume set with value ${value * 100}%`);
            break;
        default:
            logger.info('setOscillatorVolume unknown oscillatorId provided, no oscillator volume set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorGainNode = (gainNode: GainNode, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_GAIN', payload: gainNode });
            logger.info('Oscillator 1 gain node added', gainNode);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_GAIN', payload: gainNode });
            logger.info('Oscillator 2 gain node added', gainNode);
            break;
        case 'oscSub':
            dispatch({ type: 'SET_OSC_SUB_GAIN', payload: gainNode });
            logger.info('Oscillator Sub gain node added', gainNode);
            break;
        default:
            logger.info('setOscillatorGainNode unknown oscillatorId provided, no oscillator gain node set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorEnabled = (isEnabled: boolean, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_ENABLED', payload: isEnabled });
            stopOscillatorById('osc1');
            logger.info('Oscillator 1 isEnabled: ', isEnabled);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_ENABLED', payload: isEnabled });
            stopOscillatorById('osc2');
            logger.info('Oscillator 2 isEnabled: ', isEnabled);
            break;
        case 'oscSub':
            dispatch({ type: 'SET_OSC_SUB_ENABLED', payload: isEnabled });
            stopOscillatorById('oscSub');
            logger.info('Oscillator Sub isEnabled: ', isEnabled);
            break;
        default:
            logger.info('setOscillatorEnabled unknown oscillatorId provided, no oscillator set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorType = (oscillatorType: OscillatorType, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_TYPE', payload: oscillatorType });
            logger.info(`Oscillator 1 type set to ${oscillatorType}`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_TYPE', payload: oscillatorType });
            logger.info(`Oscillator 2 type set to ${oscillatorType}`);
            break;
        case 'oscSub':
            dispatch({ type: 'SET_OSC_SUB_TYPE', payload: oscillatorType });
            logger.info(`Oscillator Sub type set to ${oscillatorType}`);
            break;
        default:
            logger.info('setOscillatorType unknown oscillatorId provided, no oscillator type set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorDetune = (value: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_DETUNE', payload: value });
            logger.info(`Oscillator 1 detune set to ${value}`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_DETUNE', payload: value });
            logger.info(`Oscillator 2 detune set to ${value}`);
            break;
        default:
            logger.info('setOscillatorDetune unknown oscillatorId provided, no oscillator detune set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorPanNode = (panNode: StereoPannerNode, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_PAN_NODE', payload: panNode });
            logger.info('Oscillator 1 pan node added', panNode);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_PAN_NODE', payload: panNode });
            logger.info('Oscillator 2 pan node added', panNode);
            break;
        case 'oscSub':
            dispatch({ type: 'SET_OSC_SUB_PAN_NODE', payload: panNode });
            logger.info('Oscillator Sub pan node added', panNode);
            break;
        default:
            logger.info('setOscillatorPanNode unknown oscillatorId provided, no oscillator pan node set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorPanPosition = (position: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_PAN_POSITION', payload: position });
            logger.info('Oscillator 1 pan position set with value ', position);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_PAN_POSITION', payload: position });
            logger.info('Oscillator 2 pan position set with value ', position);
            break;
        case 'oscSub':
            dispatch({ type: 'SET_OSC_SUB_PAN_POSITION', payload: position });
            logger.info('Oscillator Sub pan position set with value ', position);
            break;
        default:
            logger.info('setOscillatorPanPosition unknown oscillatorId provided, no oscillator pan node set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorOctave = (octave: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_OCTAVE', payload: octave });
            logger.info('Oscillator 1 octave set with value ', octave);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_OCTAVE', payload: octave });
            logger.info('Oscillator 2 octave set with value ', octave);
            break;
        case 'oscSub':
            dispatch({ type: 'SET_OSC_SUB_OCTAVE', payload: octave });
            logger.info('Oscillator Sub octave set with value ', octave);
            break;
        default:
            logger.info('setOscillatorOctave unknown oscillatorId provided, no oscillator pan node set. Received: ', oscillatorId);
            break;
    }
};
