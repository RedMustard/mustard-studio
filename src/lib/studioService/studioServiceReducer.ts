import { GenericAction, StudioService } from '../../types/types';

export const studioServiceReducer = (state: StudioService, action: GenericAction) => {
    switch (action.type) {
        case 'SET_MASTER_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    master: action.payload,
                },
            };

        case 'SET_OSCILLATOR_1_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    oscillator1: action.payload,
                },
            };

        case 'SET_OSCILLATOR_2_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    oscillator2: action.payload,
                },
            };

        case 'SET_MASTER_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    master: action.payload,
                },
            };

        case 'SET_OSCILLATOR_1_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    oscillator1: action.payload,
                },
            };

        case 'SET_OSCILLATOR_2_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    oscillator2: action.payload,
                },
            };

        default:
            return state;
    }
};
