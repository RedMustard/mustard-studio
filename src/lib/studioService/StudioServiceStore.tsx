import { createContext, h } from 'preact';
import { useReducer } from 'preact/hooks';
import { OSC_1_INITIAL_SETTINGS, OSC_2_INITIAL_SETTINGS, OSC_SUB_INITIAL_SETTINGS } from '../../constants';
import { DispatchFunction, StudioService } from '../../types/types';
import { deepClone } from '../utils/objects/objects';
import { studioServiceReducer } from './studioServiceReducer';


const initialState: StudioService = Object.freeze({
    master: {
        gainNode: undefined,
        panNode: undefined,
        settings: {
            volume: 0.1,
            pan: 0,
        },
    },
    oscillators: {
        osc1: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_1_INITIAL_SETTINGS,
            },
        },
        osc2: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_2_INITIAL_SETTINGS,
            },
        },
        oscSub: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_SUB_INITIAL_SETTINGS,
            },
        },
    },
});

export const getInitialState = () => deepClone(initialState);

export const StudioServiceContext = createContext<[StudioService, DispatchFunction]>([getInitialState(), () => {}]);

export const StudioServiceStore = ({ children }: any) => {
    const [state, dispatch]: [StudioService, DispatchFunction] = useReducer(studioServiceReducer, getInitialState());

    return (
        <StudioServiceContext.Provider value={[state, dispatch]}>
            {children}
        </StudioServiceContext.Provider>
    );
};
