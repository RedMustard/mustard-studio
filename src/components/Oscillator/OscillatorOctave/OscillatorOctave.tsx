import { h } from 'preact';
import { useContext } from 'preact/hooks';
import { MAX_OSC_OCTAVE, MIN_OSC_OCTAVE, OCTAVE_STEPS } from '../../../constants';
import { resetOscillatorOctave, setOscillatorOctave } from '../../../lib/studioService/studioServiceActions';
import { StudioServiceContext } from '../../../lib/studioService/StudioServiceStore';
import { ValueUnitEnum } from '../../../types/runtimeTypes';
import { OscillatorId } from '../../../types/types';
import { Fader } from '../../Fader/Fader';

interface OscillatorOctaveProps {
    oscillatorId: OscillatorId;
}

export const OscillatorOctave = ({ oscillatorId }: OscillatorOctaveProps) => {
    const [studioService, dispatch] = useContext(StudioServiceContext);
    const { oscillators } = studioService;
    const oscillatorOctave = oscillators[oscillatorId].settings.octave;

    const handleOnMouseDown = (e: MouseEvent) => {
        if (e.buttons === 1 && (e.ctrlKey || e.metaKey)) {
            resetOscillatorOctave(oscillatorId, dispatch);
        }
    };

    return (
        <div className="oscillator-octave">
            Octave: &nbsp;
            <Fader
                classSuffix="oscillator-octave"
                value={oscillatorOctave}
                onInput={(value) => setOscillatorOctave(value, oscillatorId, dispatch)}
                onMouseDown={(e: MouseEvent) => handleOnMouseDown(e)}
                maxValue={MAX_OSC_OCTAVE}
                minValue={MIN_OSC_OCTAVE}
                stepResolution={OCTAVE_STEPS}
                valueUnit={ValueUnitEnum.CENT}
            />
        </div>
    );
};
