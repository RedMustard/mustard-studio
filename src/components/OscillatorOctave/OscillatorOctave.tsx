import { h } from 'preact';
import { useContext } from 'preact/hooks';
import { MAX_OCTAVE, MIN_OCTAVE, OCTAVE_STEPS } from '../../constants';
import { setOscillatorOctave } from '../../lib/studioService/studioServiceActions';
import { StudioServiceContext } from '../../lib/studioService/StudioServiceStore';
import { ValueUnitEnum } from '../../types/runtimeTypes';
import { OscillatorId } from '../../types/types';
import { Fader } from '../Fader/Fader';

interface OscillatorOctaveProps {
    oscillatorId: OscillatorId;
}

export const OscillatorOctave = ({ oscillatorId }: OscillatorOctaveProps) => {
    const [studioService, dispatch] = useContext(StudioServiceContext);
    const { settings } = studioService;
    const oscillatorOctave = settings[oscillatorId].octave;

    return (
        <div className="oscillator-octave">
            Octave: &nbsp;
            <Fader
                classSuffix="oscillator-octave"
                value={oscillatorOctave}
                onInput={(value) => setOscillatorOctave(value, oscillatorId, dispatch)}
                maxValue={MAX_OCTAVE}
                minValue={MIN_OCTAVE}
                stepResolution={OCTAVE_STEPS}
                valueUnit={ValueUnitEnum.CENT}
            />
        </div>
    );
};
