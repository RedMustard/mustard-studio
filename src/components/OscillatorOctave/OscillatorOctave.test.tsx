import { h } from 'preact';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { OscillatorOctave } from './OscillatorOctave';
import { setOscillatorOctave } from '../../lib/studioService/studioServiceActions';
import { OscillatorId } from '../../types/types';


jest.mock('../../lib/studioService/studioServiceActions');

// eslint-disable-next-line import/newline-after-import
const wamock = require('web-audio-mock-api');
const audioContext = new wamock.AudioContext();
const baseProps = {
    audioContext,
    oscillatorId: 'osc1' as OscillatorId,
};

describe('<OscillatorOctave />', () => {
    it('renders with basic props', () => {
        const wrapper = shallow(
            <OscillatorOctave
                {...baseProps}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('calls setOscillatorOctave when pan changed', () => {
        const wrapper = mount(
            <OscillatorOctave
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('input');
        expect(setOscillatorOctave).toBeCalled();
    });
});
