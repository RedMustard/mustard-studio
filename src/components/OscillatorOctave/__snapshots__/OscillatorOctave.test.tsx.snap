// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OscillatorOctave /> renders with basic props 1`] = `
<div
  className="oscillator-octave"
>
  Octave: Â 
  <Fader
    classSuffix="oscillator-octave"
    maxValue={8}
    minValue={-8}
    onInput={[Function]}
    stepResolution={1}
    value={0}
    valueUnit="cent"
  />
</div>
`;
